### DISCUSSION TITLE
Unresolved external symbols with jpm on Windows

### DISCUSSION THREAD

**Discussion by jangler** [janet-lang/janet#1354] [Q&A] [answered]
I'm having an issue with jpm on Windows. Here's what I've done so far:

- Install Janet via the MSI
- Add `C:\Program Files\Git\mingw64\libexec\git-core` to my PATH
- Run `bootstrap.janet`
- Install the Visual Studio C++ build tools
- Run `jpm install (package)` in Developer Command Prompt

What happens is that I get a lot of this:

```
C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools>jpm install jaylib
From https://github.com/janet-lang/jaylib
 * branch            HEAD       -> FETCH_HEAD
HEAD is now at 51e88ff Merge pull request #54 from greenfork/raylib-5.0
creating native module build/jaylib.dll...
creating static library build/jaylib.static.lib...
   Creating library build\jaylib.lib and object build\jaylib.exp
src___main.o : error LNK2019: unresolved external symbol _janet_unwrap_number referenced in function _camera2d_put
src___main.o : error LNK2019: unresolved external symbol _janet_wrap_number referenced in function _camera2d_get
src___main.o : error LNK2019: unresolved external symbol _janet_nanbox32_from_tagi referenced in function _cfun_InitWindow
... (snipped) ...
src___main.o : error LNK2019: unresolved external symbol _janet_getinteger64 referenced in function _jaylib_getcolor
src___main.o : error LNK2019: unresolved external symbol _janet_getindexed referenced in function _cfun_SetWindowIcons
src___main.o : error LNK2019: unresolved external symbol _janet_getabstract referenced in function _cfun_BeginMode2D
C:\Users\casey\AppData\Local\Apps\Janet\C\\janet.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'x86'
build\jaylib.dll : fatal error LNK1120: 41 unresolved externals
error: command failed with non-zero exit code 1120
error: build fail
  in pdag [C:/Users/casey/AppData/Local/Apps/Janet/Library/jpm/dagbuild.janet] (tailcall) on line 79, column 23
  in <anonymous> [C:/Users/casey/AppData/Local/Apps/Janet/Library/jpm/pm.janet] on line 260, column 11
  in <anonymous> [C:/Users/casey/AppData/Local/Apps/Janet/Library/jpm/pm.janet] on line 244, column 7
  in bundle-install [C:/Users/casey/AppData/Local/Apps/Janet/Library/jpm/pm.janet] on line 242, column 5
  in install [C:/Users/casey/AppData/Local/Apps/Janet/Library/jpm/commands.janet] on line 205, column 20
  in run [C:/Users/casey/AppData/Local/Apps/Janet/Library/jpm/cli.janet] (tailcall) on line 88, column 9
  in run-main [boot.janet] on line 3936, column 16
  in cli-main [boot.janet] on line 4089, column 17
```

This happens whether I use the "x86 Native Tools Command Prompt", the "x64 Native Tools Command Prompt", the "Developer Command Prompt", or whether I manually run `vcvars32.bat` or `vcvars64.bat`.

Anyone have any suggestions?

**Comment by sogaiu** [comment, marked as answer]
I don't think I've tried mixing mingw's git with using msvc things.

Typically I use git installed via scoop, which I think is a "repackaging" of [this](https://gitforwindows.org/).

May be it's worth trying that instead (whether from scoop or otherwise)?

---

By `bootstrap.janet`, do you mean a file that's in `jpm`'s repository?  I think when I do the `janet bootstrap.janet` thing, I tend to also do it via "x64 Native Tools Command Prompt" for the 64-bit-ish type of situation [1].

FWIW, I've seen reports in the past that using other command prompts can result in problems.

---

[1] Some more related details are [here](https://github.com/janet-lang/janet-lang.org/issues/192).

**Reply by jangler** [reply]
Ok, I re-ran `janet boostrap.janet` in the x64 Native Tools Command Prompt, then ran `jpm clear-cache`. Building and installing native extensions now works. Thank you!

For the record, I'm pretty sure the mingw path I mentioned *is* from Git for Windows.

**Reply by sogaiu** [reply]
Thanks for the bit about the mingw path -- I'll try to take a look at what I've got in a Windows environment the next time I have access :)
